tasks/test.py

```diff
@@ -124,37 +124,51 @@ def install_tools(ctx):
                     ctx.run(f"go install {tool}")
 
 
-# TODO(AP-1879): The following four functions all do something similar: they run a given command on a list of modules
-# This could be refactored in a core function that does the loop on modules and returns failures, and
-# wrapper functions that craft the command to run and process the results and errors.
-
-
-def lint_flavor(
-    ctx, modules: List[GoModule], flavor: AgentFlavor, build_tags: List[str], arch: str, rtloader_root: bool
+def test_core(
+    modules: List[GoModule],
+    flavor: AgentFlavor,
+    module_class: GoModule,
+    operation_name: str,
+    command,
+    skip_module_class=False,
 ):
     """
-    Runs linters for given flavor, build tags, and modules.
+    Run the command function on each module of the modules list.
     """
-    modules_lint_results = []
-    print(f"--- Flavor {flavor.name}: golangci_lint")
+    modules_results = []
+    print(f"--- Flavor {flavor.name}: {operation_name}")
     for module in modules:
-        module_lint_result = ModuleLintResult(path=module.full_path())
+        module_result = None
+        if not skip_module_class:
+            module_result = module_class(path=module.full_path())
         print(f"----- Module '{module.full_path()}'")
         if not module.condition():
             print("----- Skipped")
             continue
 
+        command(modules_results, module, module_result)
+    return modules_results
+
+
+def lint_flavor(
+    ctx, modules: List[GoModule], flavor: AgentFlavor, build_tags: List[str], arch: str, rtloader_root: bool
+):
+    """
+    Runs linters for given flavor, build tags, and modules.
+    """
+
+    def command(lint_results, module, module_result):
```

Comment:
In this method, the same `lint_results` variable is used for three distinct things, which makes its content unexpected:
- it is one of the parameters of the `command` function, which should append a new `ModuleLintResult` to it 
- however, `lint_results` is then redefined (line 162) as the output of `run_golangci_lint`, ie. a new list
- therefore, at this point, `lint_results` doesn't point to the list that was passed to the function anymore, and therefore the `lint_results.append` on line 169 doesn't really do anything.

The end result is that `lint_flavor` always returns an empty list, meaning that we are not catching linter errors.